### PR TITLE
fix(transaction): Properly store block cancel status

### PR DESCRIPTION
**Pull request by dranikpg**
This code from `RegisterBreakHook` makes sure we can even cancel ahead, so that the transaction is guaranteed to not get stuck:

```c++
 if (res->transaction)
      res->transaction->CancelBlocking(nullptr);
```

however when we do this, we can do it as early as before running the more recent callback with Execute(), which in turn overwrites local_result_

at the end we get local_result_ being set to OK, but COORD_CANCELLED being set as well

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -1412,7 +1413,8 @@ void Transaction::CancelBlocking(std::function<OpStatus(ArgSlice)> status_cb) {
     return;
 
   coordinator_state_ |= COORD_CANCELLED;
-  local_result_ = status;
+  // don't use local_result_ because it can be overwirtten if we cancel ahead
+  block_cancel_result_ = status;
```
Comment (romange) [inline src/server/transaction.cc:1417]: seems you do not need `block_cancel_result_` at all. if you reached this point, `block_cancel_result_` will be set CANCELLED, is not it?

File: src/server/transaction.cc [hunk 2]
```diff
@@ -1178,7 +1178,8 @@ OpStatus Transaction::WaitOnWatch(const time_point& tp, WaitKeysProvider wkeys_p
   if (status == cv_status::timeout) {
     result = OpStatus::TIMED_OUT;
   } else if (coordinator_state_ & COORD_CANCELLED) {
-    result = local_result_;
+    DCHECK_GT(block_cancel_result_, OpStatus::OK);
```
Comment (romange) [inline src/server/transaction.cc:1181]: won't it always be `CANCELLED` here? if not then when it's something else?

File: src/server/server_family.cc [hunk 3]
```diff
@@ -1568,7 +1568,7 @@ void ServerFamily::CancelBlockingOnThread(std::function<OpStatus(ArgSlice)> stat
   auto cb = [status_cb](unsigned thread_index, util::Connection* conn) {
     if (auto fcntx = static_cast<facade::Connection*>(conn)->cntx(); fcntx) {
       auto* cntx = static_cast<ConnectionContext*>(fcntx);
-      if (cntx->transaction && cntx->blocked) {
```
Comment (romange) [inline src/server/server_family.cc:1571]: is it also part of the fix?

**Review by romange — APPROVED** [review]

File: src/server/server_family.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/server_family.cc:1571]: Yes, that was the reason for the deadlock, we want to cancel blocking commands even in their preparation phase

**Comment by romange** [comment]
@dranikpg  what Github issues does it resolve?

**Pull request merged**